### ISSUE TITLE
Support non-modal menus or nesting of menus

### ISSUE THREAD

**Issue by ggarra13** [fltk/fltk#847]
I've been playing with blender 4.0, the 3D application and I found out they have an extremely handy feature.  They support when opening a menu and selecting an item, to use the RMB button to open another menu.  The main use is to add shortcuts to the menus.  In blender, you move the cursor to the menu item, select RMB and another menu opens which allows adding a hotkey for the menu item under the cursor.
I find this workflow extremely intuitive and would like to implement that in my FLTK application.

**I am using FLTK**
1.4

**Describe the solution you'd like**
I would like to have more control over menus to make them non-modal (ie. modify the popup() function) or have another function to add a callback to the menu with menu item (mmenu()) that would get called on RMB instead of when selected with the LMB.

**Describe alternatives you've considered**
I considered modifying popup() function in menu items, but found it too cumbersome to get rid of the modal behavior.

**Additional context**
![Captura de pantalla 2023-11-26 141133](https://github.com/fltk/fltk/assets/6072843/0c4448fc-ce6a-4f55-acc4-26b8bef96eda)
![Captura de pantalla 2023-11-26 141212](https://github.com/fltk/fltk/assets/6072843/53f8665e-0055-4af9-a4e6-0ab69d1d885f)
![Captura de pantalla 2023-11-26 141325](https://github.com/fltk/fltk/assets/6072843/c828e855-a724-40b2-96bb-23acd2fbf9fc)

**Comment by MatthiasWM** [comment]
Oh, wow. Even though I am using Blender every once in a while, I did not know about this. It should not be too difficult to add the option to open another submenu when right-clicking on a regular menu. Plus I want that in Fluid ;-)

With the current API, the right mouse button does call the callback, and you can check that using `Fl::event()` and `Fl::event_button()`. So would an API be good enough that allows a popup from within a popup? 

And yes, the menu handling code is quite complex. We have three Issues mentioning a rework in one way or another.

**Comment by ggarra13** [comment]
> With the current API, the right mouse button does call the callback, and you can check that using `Fl::event()` and `Fl::event_button()`. So would an API be good enough that allows a popup from within a popup?

I think the API would need to allow the RMB to do whatever you want, including adding a new popup.  The main issue would be that the event loop does not freeze, so you can then have the hotkey keyboard entry happening too.

**Comment by MatthiasWM** [comment]
Ok, so I tried this out. You can absolutely open another popup menu from an existing menu. The reason is simple: the menu loop dismisses the menu before it calls the callback. If it wouldn't do that, opening a dialog box from a pulldown menu would generate a recursion beeper into event loops which is not a good thing (we had this when users would open multiple dialog boxes through a timer, but this has been solved a long time ago).

In the same context, opening a pulldown from a pulldown is fine, only the first pulldown will be gone before the second pulldown opens. If there s a visual hint to the user, I don't see that as much of a problem. What do you think?

Handling the shortcut keypress would need to be done in a separate dialog window. I will implement this fully for Fluid for testing and post the code here when I am satisfied.

```
static void menu_file_open_cb(Fl_Widget *, void *) {
  if (Fl::event_button()==FL_RIGHT_MOUSE) {
    static Fl_Menu_Item mi[] = {
      { "Change shortcut" },
      { "Restore shortcut" },
      { NULL }
    };
    mi->popup(Fl::event_x(), Fl::event_y(), "File/Open");
  } else {
    open_project_file("");
  }
}
```
![Screenshot 2023-11-27 at 16 14 05](https://github.com/fltk/fltk/assets/653993/dbea252e-8800-4e80-98f1-6f9c66dd8864)
![Screenshot 2023-11-27 at 16 13 25](https://github.com/fltk/fltk/assets/653993/e9d0e20e-8284-41d0-89b1-4da82af63c49)

**Comment by ggarra13** [comment]
Indeed it works.  I have a couple of questions/issues thou.  

1) Is there a way to get the full pathname of the menu from within the callback (ie. from within Fl_Widget*, which I assume it is a Fl_Menu_Item*), so that the code can be made into a more generic function?  I don't have the pointer to the menu bar, as it can be hidden and the callback can also be called by a RMB context menu.

2) Is there a way to avoid the closing of the original menu?  Closing the original menu becomes confusing, at least to me, to know what happened.  Blender keeps the menus open and they just stack on top of each other.

**Comment by MatthiasWM** [comment]
1. Fl_Widget* in the callback is indeed an Fl_Menu_Item* and/should be cast to that. You a use that to get the label and current shortcut of the menu item. You can get the complete path using [FL_Menu_::path_name()](https://fltk.gitlab.io/fltk/classFl__Menu__.html#abe4ba98d1af637c1e452df648c19502c), but you will need the Menu Bar pointer for that call. If you know the first item in a menu array, you can fake this call. This has been bugging me forever ;-) .

2. No, not with the current code. If we keep the menu windows open, you would be very limited to what you do during the callback (basically limited to opening other menu windows). But that is not implemented either. I like what Blender did there, and I will try to add something to Fluid that is isolated enough for users to reuse the code in their apps. If it works well, it may make its way into the core library eventually.

**Comment by ggarra13** [comment]
The other problem I am facing when implementing this is internationalization.  The menus, depending of the language chosen, will be in a different language.  So, storing a mapping for the shortcut, like "File/Open" -> CTRL+o, will work only as long as the user is using my program in English.
In my application, currently the problem is solved as my calls are like:

```
        menu->add(
            _("File/Open/Movie or Sequence"), kOpenImage.hotkey(),
            (Fl_Callback*)open_cb, ui);
```

where kOpenImage is actually a class that stores the hotkey and returns it with the hotkey() call.  And the hotkey preferences always store File/Open/... -> kOpenImage.hotkey() in English.  The menus are recreated after every complex operation on them.  

If the change of the hotkey is done from the callback as you suggested, I *can* work around it, by changing kOpenImage directly, but I thought I should mention this so you take this into account when implementing your solution.

Also... BTW, completely unrelated, I posted a new message to the Fl_Tile change you did in the thread about new sizes Issue #797 .  When I tried it, it seemed like the hidden group/box functionality was no longer working.  I am busy at the moment.  Can you verify it works for you with Albretch's mrViewer1.cxx from that thread?

**Comment by MatthiasWM** [comment]
> The other problem I am facing when implementing this is internationalization. 

Yeah, I think you may have to keep the original menu around and make a copy and assign that to Fl_Menubar and other classes based on Fl_Menu_. You can then find the menu item using the English path, convert the menu item pointer into an index, and then convert the index back into the menu item pointer in the current language. All this functions (copy, ptr to index and index to ptr) exists in Fl_Menu_ IIRC. I am native German, so internationalization is a thing here too.

> Also... BTW, completely unrelated, I posted a new message to the Fl_Tile change you did in the thread about new sizes Issue #797 . When I tried it, it seemed like the hidden group/box functionality was no longer working. I am busy at the moment. Can you verify it works for you with Albretch's mrViewer1.cxx from that thread?

Yeah, I check that. What I can say already is, that Fl_Tile naw has two modes, and if you use any of the size range related calls, the resizable() widget will get more of its original meaning back. There are some brief docs: https://fltk.gitlab.io/fltk/classFl__Tile.html#details

**Comment by MatthiasWM** [comment]
See #768: I will probably not be able to implement this before feature freeze, and TBH, this should go with the pulldown menu rework. A partial solution is already in the 4th message, so it can be done with the current API.
